File: examples/rop.cpp
#include <exception>
#include <iostream>
#include <string_view>
#include <xcraft/xcraft.hpp>

int main() {
    auto rop = xcft::ROP("./bin/vuln");
    try {
        auto gadgets = rop.gadgets();
        for (auto g : gadgets) {
            std::cout << "0x" << std::hex << std::setw(16) << std::setfill('0')
                      << g.ins[0].address;
            for (auto i : g.ins)
                std::cout << " " << i.assembly << ";";
            std::cout << std::endl;
        }

        auto addresses = rop.find_gadget({"pop rdi", "ret"});
        for (auto a : addresses)
            std::cout << "0x" << std::hex << std::setw(16) << std::setfill('0')
                      << a << std::endl;
    } catch (const std::runtime_error &e) {
        std::cerr << e.what() << std::endl;
    }
}

File: examples/rop.py
from pwn import *

context.binary = './bin/vuln'         
elf = context.binary
rop = ROP(elf)

for addr, g in sorted(rop.gadgets.items()):
    print(f"0x{addr:016x}", end='')
    for insn in g.insns:
        print(f" {insn};", end='')
    print()

g = rop.find_gadget(['pop rdi', 'ret'])
if g:
    print(f"0x{g.address:016x}")

matches = [addr for addr, g in rop.gadgets.items() if g.insns == ['pop rdi', 'ret']]
for a in sorted(matches):
    print(f"0x{a:016x}")